### PR TITLE
fix(ngcc): prevent including JavaScript sources outside of the package

### LINKED ISSUE #37508: Error on worker #5: Error: Expected UMD module file (Angular 9.1.9)
# 🐞 bug report

### Description

After upgrading my SSR application to angular from 9.1.0 to 9.1.9, I got the following error when I build the application on SSR using `npm run build:ssr`

## 🔥 Exception or Error
<pre>

Error: Error on worker #5: Error: Expected UMD module file (G:/workspace/project/node_modules/flat/index.js) to contain exactly one statement, but found 6.
    at UmdReflectionHost.computeUmdModule (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:75:23)
    at FactoryMap.factory (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:31:84)
    at FactoryMap.get (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\utils.js:93:48)
    at UmdReflectionHost.getUmdModule (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:56:36)
    at UmdReflectionHost.getModuleStatements (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:70:34)
    at UmdReflectionHost.computeExportsOfUmdModule (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:84:53)
    at FactoryMap.factory (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:32:84)
    at FactoryMap.get (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\utils.js:93:48)
    at UmdReflectionHost.getExportsOfModule (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\umd_host.js:50:94)
    at DelegatingReflectionHost.getExportsOfModule (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\host\delegating_host.js:71:34)
    at ClusterMaster.onWorkerMessage (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\execution\cluster\master.js:194:27)
    at G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\execution\cluster\master.js:54:95
    at ClusterMaster.<anonymous> (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\execution\cluster\master.js:292:57)
    at step (G:\workspace\project\node_modules\tslib\tslib.js:139:27)
    at Object.next (G:\workspace\project\node_modules\tslib\tslib.js:120:57)
    at G:\workspace\project\node_modules\tslib\tslib.js:113:75
    at new Promise (<anonymous>)
    at Object.__awaiter (G:\workspace\project\node_modules\tslib\tslib.js:109:16)
    at EventEmitter.<anonymous> (G:\workspace\project\node_modules\@angular\compiler-cli\ngcc\src\execution\cluster\master.js:286:32)
    at EventEmitter.emit (events.js:198:13)<code>

</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre>
Angular CLI: 9.1.0
Node: 10.16.2
OS: win32 x64

Angular: 9.1.0
... animations, cli, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.7
@angular-devkit/build-angular     0.901.7
@angular-devkit/build-optimizer   0.901.7
@angular-devkit/build-webpack     0.901.7
@angular-devkit/core              9.1.0
@angular-devkit/schematics        9.1.0
@angular/cdk                      9.2.0
@angular/fire                     6.0.0
@angular/flex-layout              9.0.0-beta.29
@angular/material                 9.2.0
@ngtools/webpack                  9.1.7
@nguniversal/builders             9.1.0
@nguniversal/common               9.0.2
@nguniversal/express-engine       9.0.2
@schematics/angular               9.1.0
@schematics/update                0.901.0
rxjs                              6.5.5
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

## Dependencies : package.json

<pre>
  "dependencies": {
    "@angular-ru/ngx-i18n-combine": "^1.1.0",
    "@angular/animations": "~9.1.9",
    "@angular/cdk": "^9.2.0",
    "@angular/common": "~9.1.9",
    "@angular/compiler": "~9.1.9",
    "@angular/core": "~9.1.9",
    "@angular/fire": "^6.0.0",
    "@angular/flex-layout": "^9.0.0-beta.29",
    "@angular/forms": "~9.1.9",
    "@angular/material": "^9.2.0",
    "@angular/platform-browser": "~9.1.9",
    "@angular/platform-browser-dynamic": "~9.1.9",
    "@angular/platform-server": "~9.1.9",
    "@angular/router": "~9.1.9",
    "@briebug/cypress-schematic": "^3.3.0",
    "@nestjs/common": "^6.0.0",
    "@nestjs/core": "^6.0.0",
    "@nestjs/ng-universal": "^3.1.0",
    "@nestjs/platform-express": "^6.0.0",
    "@ngneat/transloco": "^2.14.0",
    "@nguniversal/express-engine": "^9.0.2",
    "@sentry/browser": "^5.15.4",
    "@squadette/hammerjs": "2.1.0-pre3",
    "class-transformer": "^0.2.3",
    "class-validator": "^0.9.1",
    "express": "^4.15.2",
    "firebase": "^7.13.2",
    "reflect-metadata": "^0.1.13",
    "rxjs": "^6.5.5",
    "tslib": "^1.10.0",
    "zone.js": "~0.10.2"
  },
  "devDependencies": {
    "@angular-builders/jest": "^9.0.1",
    "@angular-devkit/architect": "~0.900",
    "@angular-devkit/build-angular": "~0.901.7",
    "@angular/cli": "~9.1.7",
    "@angular/compiler-cli": "~9.1.9",
    "@angular/language-service": "~9.1.9",
    "@angularclass/hmr": "^2.1.3",
    "@cypress/webpack-preprocessor": "5.1.1",
    "@nguniversal/builders": "^9.0.2",
    "@types/express": "^4.17.0",
    "@types/jest": "^25.2.1",
    "@types/node": "^12.11.1",
    "codelyzer": "^5.1.2",
    "cypress": "4.3.0",
    "firebase-tools": "^8.0.0",
    "fuzzy": "^0.1.3",
    "husky": "3.1.0",
    "inquirer": "^6.2.2",
    "inquirer-autocomplete-prompt": "^1.0.1",
    "jest": "^25.2.7",
    "jest-junit": "^10.0.0",
    "prettier": "^2.0.4",
    "pretty-quick": "^2.0.1",
    "replace-in-file": "^6.0.0",
    "rxjs-tslint-rules": "^4.33.0",
    "ts-jest": "^25.3.1",
    "ts-loader": "6.2.2",
    "ts-node": "~8.3.0",
    "tslint": "~6.1.0",
    "tslint-plugin-prettier": "^2.3.0",
    "typescript": "~3.8.3",
    "typescript-tslint-plugin": "^0.5.5"
  },
</code></pre>

### PR DESCRIPTION
**Pull request by JoostK**
See individual commits for details.

Fixes #37508

### CODE REVIEW

**Review by petebacondarwin — APPROVED** [review]
Couple of nits.
Nice and clean PR!

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 1]
```diff
@@ -401,6 +401,127 @@ runInEachFileSystem(() => {
...
+           },
+           {
+             name: _('/node_modules/test-package/index.js'),
+             contents: `
+            (function (global, factory) {
+              typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
+              typeof define === 'function' && define.amd ? define('test', ['exports'], factory) :
+              (factory(global.test));
+            }(this, (function (exports) { 'use strict';
+              var identity_1 = require("identity");
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:446]: I got this all wrong, haha. The whole `require` business should be in the UMD wrapper itself.

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 2]
```diff
@@ -401,6 +401,127 @@ runInEachFileSystem(() => {
...
+           },
+         ]);
+
+         compileIntoApf('test-package', {
+           '/index.ts': `
+          import {NgModule, Component} from '@angular/core';
+          import {selector} from 'external';
+
+          @Component({
+            selector: selector,
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:497]: Sure

File: packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts [hunk 3]
```diff
@@ -401,6 +401,127 @@ runInEachFileSystem(() => {
...
+              (factory(global.test));
+            }(this, (function (exports) { 'use strict';
+              var identity_1 = require("identity");
+              var Foo = /** @class */ (function () {
+                  function Foo() {
+                  }
+                  Foo.doSomething = function (x) {
+                      // The below statement is analyzed for 'ModuleWithProviders', so is evaluated
+                      // by ngcc. The reference into the non-TS CommonJS package used to crash ngcc.
+                      return identity_1.identity(x);
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/test/integration/ngcc_spec.ts:453]: Correct, changed.

File: packages/compiler-cli/ngcc/test/helpers/utils.ts [hunk 4]
```diff
@@ -68,7 +68,7 @@ export function makeTestBundleProgram(
   const rootDir = fs.dirname(entryPointPath);
   const options: ts.CompilerOptions =
       {allowJs: true, maxNodeModuleJsDepth: Infinity, checkJs: false, rootDir, rootDirs: [rootDir]};
-  const host = new NgccSourcesCompilerHost(fs, options, entryPointPath);
+  const host = new NgccSourcesCompilerHost(fs, options, rootDir);
```
[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/ngcc/test/helpers/utils.ts:71]: This code has always made the assumption that `path` corresponds with a secondary entry-point, while that is not true in most cases. Cleaning up would likely be a large task, given that this function has ~480 callers. I could add an optional param here but that doesn't sounds like a proper way to go either.

**Comment by JoostK** [comment]
merge-assistance: angular bot seems to have stalled, all checks should have passed.

**Comment by matheo** [comment]
@JoostK help please!
We're upgrading from Angular v8 to v9
and our libraries are aliasing the exports with the company prefix.
So `VendorService` is exported as `PrefixedVendorService`.

I'm trying to upgrade a website to `v9` keeping the Angular `v8` libraries built with ng-packagr.
but Ivy is throwing this kind of exception while building the website for production:
```
ERROR in ./node_modules/@company/communication/__ivy_ngcc__/fesm2015/app.js 5011:315-335
"export 'VendorService' (imported as 'ɵngcc7') was not found in '@company/services'
    at HarmonyImportSpecifierDependency._getErrors (./node_modules/webpack/lib/dependencies/HarmonyImportSpecifierDependency.js:109:11)
    at HarmonyImportSpecifierDependency.getErrors (./node_modules/webpack/lib/dependencies/HarmonyImportSpecifierDependency.js:68:16)
    at Compilation.reportDependencyErrorsAndWarnings (./node_modules/webpack/lib/Compilation.js:1463:22)
    at hooks.finishModules.callAsync.err (./node_modules/webpack/lib/Compilation.js:1258:10)
    at AsyncSeriesHook.eval [as callAsync] (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:33:1)
    at AsyncSeriesHook.lazyCompileHook (./node_modules/tapable/lib/Hook.js:154:20)
    at Compilation.finish (./node_modules/webpack/lib/Compilation.js:1253:28)
    at hooks.make.callAsync.err (./node_modules/webpack/lib/Compiler.js:672:17)
    at _done (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:9:1)
    at _err1 (eval at create (./node_modules/tapable/lib/HookCodeFactory.js:33:10), <anonymous>:36:22)
    at _addModuleChain (./node_modules/webpack/lib/Compilation.js:1185:12)
    at processModuleDependencies.err (./node_modules/webpack/lib/Compilation.js:1097:9)
    at process._tickCallback (internal/process/next_tick.js:61:11)
 @ ./src/app/applications/applications.module.ts
 @ ./src/app/app.module.ts
 @ ./src/main.ts
 @ multi ./src/main.ts
```
changing the target of the libraries from `es5` to `es2015` thrown a different error mentioning the missing generic of `ModuleWithProviders`. I used `<any>` and this exception came back.

 I've also used this postinstall script with no joy:
`ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points`

Do you have any clue to share with us? or we should disable Ivy in the meantime? :(

**Comment by matheo** [comment]
Reading #32902 I guess that something is happening with the definition files
because I see the aliases at `node_modules/@company/services/index.d.ts`
but Ivy keeps looking for the unprefixed name :(

**Comment by JoostK** [comment]
@matheo I can't tell from the top of my head, sorry. Aliased exports are generally supported I think. For some reason, compiling `@company/communication` introduced an import of the internal name `VendorService` from `@company/services`, whereas you mention this has only been exported as using a prefix. I briefly looked into how that could happen but I was unable to determine a potential path, so it's a bit hard offer any advice.

Regarding the `ModuleWithProviders` error, the generic type has only been made required in Angular 10, and your version report shows that you're using Angular 9. It could be that the Angular compiled itself required the generic type and produced an error, however ngcc should have added it automatically for libraries. Note that using `any` as the generic type arg is definitely an issue, as it **must be** the NgModule type that should be imported.

Without a reproduction I'm afraid I can't really give more accurate and actionable advice :-( We have not seen issues with private declarations together with ngcc in a very long time, so I'm really not sure what is wrong here.

**Comment by matheo** [comment]
Thanks for the answer @JoostK 
I used this PR as excuse because I wanted your opinion :) 

I will trace the problem a bit and open a new issue to discuss this particular case.
Thanks again for your time!

**Pull request closed without merging**